File: Ohtap E Commerce Domain Models/Development/NetFramework/Models Core/Models Core/System/List.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace Ohtap.ECommerce.Domain.Models.Core.System
{
    public class Lista<T> : ILista<T>
    {
        private List<T> innerList = new List<T>();


        public Lista()
        {

        }

        public T this[int index] { get { return this.innerList[index]; } set { this.innerList[index] = value; } }

        public int Count { set { } get { return this.innerList.Count; } }

        public bool IsReadOnly { set { } get { return false; } }

        public void Add(T item)
        {
            this.innerList.Add(item);
        }

        public void Clear()
        {
            this.innerList.Clear();
        }

        public bool Contains(T item)
        {
            return this.innerList.Contains(item);
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            this.innerList.CopyTo(array, arrayIndex);
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.innerList.GetEnumerator();
        }

        public int IndexOf(T item)
        {
            return this.innerList.IndexOf(item);
        }

        public void Insert(int index, T item)
        {
            this.innerList.Insert(index, item);
        }

        public bool Remove(T item)
        {
            return this.innerList.Remove(item);
        }

        public void RemoveAt(int index)
        {
            this.innerList.RemoveAt(index);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.innerList.GetEnumerator();
        }
    }
}

File: Ohtap E Commerce Domain Models/Development/PHP/Core/System/List.php
<?php
namespace Ohtap\ECommerce\DomainModels\Core\System;

/*
 * https://medium.com/@moinuddinchowdhury/serialize-vs-json-67fe872a7755
 * https://www.php.net/manual/tr/class.ds-collection.php
 *https://stackoverflow.com/questions/59905261/how-do-you-properly-clear-a-json-encode-error-state
 */

class Lista extends \ArrayObject implements \IteratorAggregate, \JsonSerializable  {
    public $innerList;

    public function __construct($input = array(), $flags = 0, $iterator_class = "ArrayIterator")
    {
        parent::__construct($input, $flags, $iterator_class);
        $this->innerList = array();
    }

    public function Count(){
        return count($this->innerList);
    }
    public function Add($item){
        $this->innerList[] = $item;
    }
    public function Insert($index, $item){
        $this->innerList[$index] = $item;
    }
    public function Clear(){
        unset($this->innerList);
    }
    public function IndexOf($item){
        return array_search($this->innerList, $item);
    }
    public function Remove($item){
        $index =  $this->IndexOf($item);
        unset($this->innerList[$index]);
    }
    public function RemoveAt($index, $item){
        unset($this->innerList[$index]);
    }


    public function __get($name)
    {
        return $this->innerList[$name];
    }
    public function __set($name, $value)
    {
        return $this->innerList[$name] = $value;
    }
    public function __isset($name)
    {
        return isset($this->innerList);
    }
    public function __unset($name)
    {
        unset($this->innerList);
    }


    public function getIterator() {
        return new \ArrayIterator($this->innerList);
    }

    public function toArray() : array {
        return $this->innerList;
    }


    public function offsetSet($index, $newval)
    {
//        if (!is_int($newval)) {
//            throw new \InvalidArgumentException("Must be int");
//        }

        parent::offsetSet($index, $newval);
    }

    public function jsonSerialize()
    {
        return $this->innerList;
    }
}
